Sources/Types.swift

```diff
@@ -34,15 +34,73 @@ public typealias GetState = () -> State
+public typealias AsyncProvider = Executor
+
+/// Entity capable of executing a closure. This can be useful in tests to control where and when tasks
+/// are executed.
+///
+/// - warning: Katana assumes that implementations of this protocol will always guarantee that the tasks
+/// given as arguments are never executed in parallel.
+public protocol Executor {
+  /// Calling this method will enqueue a task to be executed asynchronously
+  @available(*, deprecated, message: "Use the new `executeAsync` instead")
   func execute(_ closure: @escaping () -> Void)
+
+  /// Calling this method will immediately execute a task, blocking the current thread until the task is
+  /// finished. It needs to be reentrant (it can be called again from the closure argument without
+  /// causing a deadlock).
+  @discardableResult
+  func executeSync<T>(_ closure: @escaping () -> T) -> T
 }
 
-/// DispatchQueue conformance to `AsyncProvider` so that `DispatchQueue.main` can be used
-extension DispatchQueue: AsyncProvider {
+extension Executor {
+  /// Calling this method will enqueue a task to be executed asynchronously
+  ///
+  /// Note: this is the method that supersedes `execute(_:)`
+  public func executeAsync(_ closure: @escaping () -> Void) {
+    self.execute(closure)
+  }
+
+  /// Calling this method will immediately execute a task, blocking the current thread until the task is
+  /// finished
+  ///
+  /// This is an inefficient default implementation made to avoid having a breaking change in the code. It
+  /// will be removed in the future.
+  @available(*, deprecated, message: "Use a proper primitive given by your underlying implementation")
+  public func executeSync<T>(_ closure: @escaping () -> T) -> T {
+    let promise = Promise<T>()
+
+    self.executeAsync {
+      promise.resolve(closure())
+    }
+
+    return try! Hydra.await(promise) // swiftlint:disable:this force_try
+  }
+}
+
+/// DispatchQueue conformance to `Executor` so that `DispatchQueue.main` can be used
+extension DispatchQueue: Executor {
   public func execute(_ closure: @escaping () -> Void) {
     self.async(execute: closure)
   }
+
+  public func executeSync<T>(_ closure: @escaping () -> T) -> T {
+    // Since `DispatchQueue.sync` is not reentrant (it will just crash the app) a proper
+    // check is needed. The idea is to use `.setSpecific` to add a value to the queue, and
+    // then `.getSpecific` to get the value back. Note that `.getSpecific` does not operate
+    // on `self`, but on the queue this thread is belonging to. This means that `.getSpecific`
+    // will return the value set only if this function is executed in the queue (and therefore,
+    // the closure must not be wrapped in a `self.sync`).
+
+    let key = DispatchSpecificKey<Void>()
+    self.setSpecific(key: key, value: ())
+
+    if DispatchQueue.getSpecific(key: key) != nil {
+      return closure()
+    } else {
+      return self.sync(execute: closure)
+    }
```

Comment:
Shouldn't the key depend on the Queue (or the key be cleared afterwards)? otherwise, if I do an executeSync from any DispatchQueue which has done an executeSync, the key will be found and the closure executed on the current dispatchqueue instead of `self`

Reply:
i'd rather clear the key afterward.